grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -51,14 +60,546 @@ class ApiGetOpenApiDescriptionHandler(api_call_handler_base.ApiCallHandler):
 
+  def __init__(self, router):
+    self.router = router
+    self.open_api_obj = None # The main OpenAPI description object.
+    self.schema_objs = None
+
+    self.proto_primitive_types_names = {
+      protobuf2.TYPE_DOUBLE: "protobuf2.TYPE_DOUBLE",
+      protobuf2.TYPE_FLOAT: "protobuf2.TYPE_FLOAT",
+      protobuf2.TYPE_INT64: "protobuf2.TYPE_INT64",
+      protobuf2.TYPE_UINT64: "protobuf2.TYPE_UINT64",
+      protobuf2.TYPE_INT32: "protobuf2.TYPE_INT32",
+      protobuf2.TYPE_FIXED64: "protobuf2.TYPE_FIXED64",
+      protobuf2.TYPE_FIXED32: "protobuf2.TYPE_FIXED32",
+      protobuf2.TYPE_BOOL: "protobuf2.TYPE_BOOL",
+      protobuf2.TYPE_STRING: "protobuf2.TYPE_STRING",
+      protobuf2.TYPE_BYTES: "protobuf2.TYPE_BYTES",
+      protobuf2.TYPE_UINT32: "protobuf2.TYPE_UINT32",
+      protobuf2.TYPE_SFIXED32: "protobuf2.TYPE_SFIXED32",
+      protobuf2.TYPE_SFIXED64: "protobuf2.TYPE_SFIXED64",
+      protobuf2.TYPE_SINT32: "protobuf2.TYPE_SINT32",
+      protobuf2.TYPE_SINT64: "protobuf2.TYPE_SINT64",
+    }
+    self.primitive_types_names = \
+      list(self.proto_primitive_types_names.values()) \
+      + ["BinaryStream",]
+
+  def _SimplifyPathNode(self, node: str) -> str:
+    if len(node) > 0 and node[0] == '<' and node[-1] == '>':
+      node = node[1:-1]
+      node = node.split(":")[-1]
+      node = f"{{{node}}}"
+
+    return node
+
+  def _SimplifyPath(self, path: str) -> str:
+    """Keep only fixed parts and parameter names from Werkzeug URL patterns.
+
+    The OpenAPI specification requires that parameters are surrounded by { }
+    which are added in _SimplifyPathNode.
+    """
+
+    nodes = path.split("/")
+    simple_nodes = [self._SimplifyPathNode(node) for node in nodes]
+
+    simple_path = '/'.join(simple_nodes)
+
+    return simple_path
+
+  def _GetPathArgsFromPath(self, path: str) -> [str]:
+    """Extract path parameters from a Werkzeug Rule URL."""
+    path_args = []
+
+    nodes = path.split("/")
+    for node in nodes:
+      if len(node) > 0 and node[0] == '<' and node[-1] == '>':
+        simple_node = self._SimplifyPathNode(node)
+        simple_node = simple_node[1:-1]
+        path_args.append(simple_node)
+
+    return path_args
+
+  def _GetTypeName(self, t):
+    if isinstance(t, FieldDescriptor):
+      if t.message_type:
+        return self._GetTypeName(t.message_type)
+      if t.enum_type:
+        return self._GetTypeName(t.enum_type)
+
+      return self._GetTypeName(t.type)
+
+    if isinstance(t, Descriptor):
+      return t.full_name
+
+    if isinstance(t, EnumDescriptor):
+      return t.full_name
+
+    if inspect.isclass(t):
+      return t.__name__
+
+    if isinstance(t, int): # It's a protobuf.Descriptor.type value.
+      return self.proto_primitive_types_names[t]
+
+    return str(t) # Cover "BinaryStream" and None.
+
+  def _SetMetadata(self):
+    oas_version = "3.0.3"
+    self.open_api_obj["openapi"] = oas_version
+
+    # The Info Object "info" field.
+    info_obj = dict()
```

Comment:
Could this be less verbose with a dict notation? (All occurrences in this file?)

```
info = {
 "contact": {
   "name": ...
 }
```
}

EDIT: The code is indeed to verbose, and would benefit from more ```{"foo":"bar"}```

Reply:
Thank you for this suggestion! I made the conversion to the `{}` dictionary notation wherever I thought it possible in the current form of the code. I also believe that the approach for representing the structure of the root OpenAPI object while building it, which you suggested in the top-level comment, is very elegant and makes the structure much easier to understand. I am currently working on making the required refactoring for this. Thank you very much!